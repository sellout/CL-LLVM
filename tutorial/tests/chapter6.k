extern printd(x);

def binary : 1 (x y) 0;  # Low-precedence operator that ignores operands.

printd(123) : printd(456) : printd(789);

# Logical unary not.
def unary!(v)
  if v then
    0
  else
    1;
    
# Unary negate.
def unary-(v)
  0-v;

# Define > with the same precedence as <.
def binary> 10 (LHS RHS)
  RHS < LHS;

# Binary logical or, which does not short circuit. 
def binary| 5 (LHS RHS)
  if LHS then
    1
  else if RHS then
    1
  else
    0;

# Binary logical and, which does not short circuit. 
def binary& 6 (LHS RHS)
  if !LHS then
    0
  else
    !!RHS;

# Define = with slightly lower precedence than relationals.
def binary = 9 (LHS RHS)
  !(LHS < RHS | LHS > RHS);

extern putchard(char)
def printdensity(d)
  if d > 8 then
    putchard(32)  # ' '
  else if d > 4 then
    putchard(46)  # '.'
  else if d > 2 then
    putchard(43)  # '+'
  else
    putchard(42); # '*'

printdensity(1): printdensity(2): printdensity(3) : 
    printdensity(4): printdensity(5): printdensity(9): putchard(10);

# determine whether the specific location diverges.
# Solve for z = z^2 + c in the complex plane.
def mandleconverger(real imag iters creal cimag)
  if iters > 255 | (real*real + imag*imag > 4) then
    iters
  else
    mandleconverger(real*real - imag*imag + creal,
                    2*real*imag + cimag,
                    iters+1, creal, cimag);

# return the number of iterations required for the iteration to escape
def mandleconverge(real imag)
  mandleconverger(real, imag, 0, real, imag);

# compute and plot the mandlebrot set with the specified 2 dimensional range
# info.
def mandelhelp(xmin xmax xstep   ymin ymax ystep)
  for y = ymin, y < ymax, ystep in (
    (for x = xmin, x < xmax, xstep in
       printdensity(mandleconverge(x,y)))
    : putchard(10)
  )
 
# mandel - This is a convenient helper function for ploting the mandelbrot set
# from the specified position with the specified Magnification.
def mandel(realstart imagstart realmag imagmag) 
  mandelhelp(realstart, realstart+realmag*78, realmag,
             imagstart, imagstart+imagmag*40, imagmag);

mandel(-2.3, -1.3, 0.05, 0.07);
mandel(-2, -1, 0.02, 0.04);
mandel(-0.9, -1.4, 0.02, 0.03);

