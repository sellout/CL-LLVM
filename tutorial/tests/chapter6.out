ready> 
ready> Read extern: 
declare double @printd(double)


ready> 
ready> Read function definition:
define double @"binary:"(double %x, double %y) {
entry:
  ret double 0.000000e+00
}


ready> 
ready> Evaluated to 0.0D0
ready> 
ready> Read function definition:
define double @"unary!"(double %v) {
entry:
  %ifcond = fcmp ueq double %v, 0.000000e+00
  %iftmp = select i1 %ifcond, double 1.000000e+00, double 0.000000e+00
  ret double %iftmp
}


ready> 
ready> Read function definition:
define double @unary-(double %v) {
entry:
  %subtmp = fsub double 0.000000e+00, %v
  ret double %subtmp
}


ready> 
ready> Read function definition:
define double @"binary>"(double %LHS, double %RHS) {
entry:
  %cmptmp = fcmp ult double %RHS, %LHS
  %booltmp = uitofp i1 %cmptmp to double
  ret double %booltmp
}


ready> 
ready> Read function definition:
define double @"binary|"(double %LHS, double %RHS) {
entry:
  %ifcond = fcmp ueq double %LHS, 0.000000e+00
  br i1 %ifcond, label %else, label %ifcont

else:                                             ; preds = %entry
  %ifcond1 = fcmp ueq double %RHS, 0.000000e+00
  %iftmp = select i1 %ifcond1, double 0.000000e+00, double 1.000000e+00
  ret double %iftmp

ifcont:                                           ; preds = %entry
  ret double 1.000000e+00
}


ready> 
ready> Read function definition:
define double @"binary&"(double %LHS, double %RHS) {
entry:
  %unop = call double @"unary!"(double %LHS)
  %ifcond = fcmp ueq double %unop, 0.000000e+00
  br i1 %ifcond, label %else, label %ifcont

else:                                             ; preds = %entry
  %unop1 = call double @"unary!"(double %RHS)
  %unop2 = call double @"unary!"(double %unop1)
  ret double %unop2

ifcont:                                           ; preds = %entry
  ret double 0.000000e+00
}


ready> 
ready> Read function definition:
define double @"binary="(double %LHS, double %RHS) {
entry:
  %cmptmp = fcmp ult double %LHS, %RHS
  %booltmp = uitofp i1 %cmptmp to double
  %binop = call double @"binary>"(double %LHS, double %RHS)
  %binop1 = call double @"binary|"(double %booltmp, double %binop)
  %unop = call double @"unary!"(double %binop1)
  ret double %unop
}


ready> 
ready> Read extern: 
declare double @putchard(double)


ready> Read function definition:
define double @printdensity(double %d) {
entry:
  %binop = call double @"binary>"(double %d, double 8.000000e+00)
  %ifcond = fcmp ueq double %binop, 0.000000e+00
  br i1 %ifcond, label %else, label %then

then:                                             ; preds = %entry
  %calltmp = call double @putchard(double 3.200000e+01)
  ret double %calltmp

else:                                             ; preds = %entry
  %binop1 = call double @"binary>"(double %d, double 4.000000e+00)
  %ifcond2 = fcmp ueq double %binop1, 0.000000e+00
  br i1 %ifcond2, label %else4, label %then3

then3:                                            ; preds = %else
  %calltmp6 = call double @putchard(double 4.600000e+01)
  ret double %calltmp6

else4:                                            ; preds = %else
  %binop7 = call double @"binary>"(double %d, double 2.000000e+00)
  %ifcond8 = fcmp ueq double %binop7, 0.000000e+00
  br i1 %ifcond8, label %else10, label %then9

then9:                                            ; preds = %else4
  %calltmp12 = call double @putchard(double 4.300000e+01)
  ret double %calltmp12

else10:                                           ; preds = %else4
  %calltmp13 = call double @putchard(double 4.200000e+01)
  ret double %calltmp13
}


ready> 
ready> Evaluated to 0.0D0
ready> 
ready> Read function definition:
define double @mandleconverger(double %real, double %imag, double %iters, double %creal, double %cimag) {
entry:
  %binop = call double @"binary>"(double %iters, double 2.550000e+02)
  %multmp = fmul double %real, %real
  %multmp1 = fmul double %imag, %imag
  %addtmp = fadd double %multmp, %multmp1
  %binop2 = call double @"binary>"(double %addtmp, double 4.000000e+00)
  %binop3 = call double @"binary|"(double %binop, double %binop2)
  %ifcond = fcmp ueq double %binop3, 0.000000e+00
  br i1 %ifcond, label %else, label %ifcont

else:                                             ; preds = %entry
  %subtmp = fsub double %multmp, %multmp1
  %addtmp6 = fadd double %subtmp, %creal
  %multmp7 = fmul double %real, 2.000000e+00
  %multmp8 = fmul double %multmp7, %imag
  %addtmp9 = fadd double %multmp8, %cimag
  %addtmp10 = fadd double %iters, 1.000000e+00
  %calltmp = call double @mandleconverger(double %addtmp6, double %addtmp9, double %addtmp10, double %creal, double %cimag)
  ret double %calltmp

ifcont:                                           ; preds = %entry
  ret double %iters
}


ready> 
ready> Read function definition:
define double @mandleconverge(double %real, double %imag) {
entry:
  %calltmp = call double @mandleconverger(double %real, double %imag, double 0.000000e+00, double %real, double %imag)
  ret double %calltmp
}


ready> 
ready> Read function definition:
define double @mandelhelp(double %xmin, double %xmax, double %xstep, double %ymin, double %ymax, double %ystep) {
entry:
  br label %loop

loop:                                             ; preds = %afterloop, %entry
  %y = phi double [ %ymin, %entry ], [ %nextvar4, %afterloop ]
  br label %loop1

loop1:                                            ; preds = %loop1, %loop
  %x = phi double [ %xmin, %loop ], [ %nextvar, %loop1 ]
  %calltmp = call double @mandleconverge(double %x, double %y)
  %calltmp2 = call double @printdensity(double %calltmp)
  %nextvar = fadd double %x, %xstep
  %cmptmp = fcmp ult double %x, %xmax
  br i1 %cmptmp, label %loop1, label %afterloop

afterloop:                                        ; preds = %loop1
  %calltmp3 = call double @putchard(double 1.000000e+01)
  %binop = call double @"binary:"(double 0.000000e+00, double %calltmp3)
  %nextvar4 = fadd double %y, %ystep
  %cmptmp5 = fcmp ult double %y, %ymax
  br i1 %cmptmp5, label %loop, label %afterloop8

afterloop8:                                       ; preds = %afterloop
  ret double 0.000000e+00
}


ready> Read function definition:
define double @mandel(double %realstart, double %imagstart, double %realmag, double %imagmag) {
entry:
  %multmp = fmul double %realmag, 7.800000e+01
  %addtmp = fadd double %multmp, %realstart
  %multmp1 = fmul double %imagmag, 4.000000e+01
  %addtmp2 = fadd double %multmp1, %imagstart
  %calltmp = call double @mandelhelp(double %realstart, double %addtmp, double %realmag, double %imagstart, double %addtmp2, double %imagmag)
  ret double %calltmp
}

123.000000
456.000000
789.000000
**++. 
*******************************************************************************
*******************************************************************************
****************************************++++++*********************************
************************************+++++...++++++*****************************
*********************************++++++++.. ...+++++***************************
*******************************++++++++++..   ..+++++**************************
******************************++++++++++.     ..++++++*************************
****************************+++++++++....      ..++++++************************
**************************++++++++.......      .....++++***********************
*************************++++++++.   .            ... .++**********************
***********************++++++++...                     ++**********************
*********************+++++++++....                    .+++*********************
******************+++..+++++....                      ..+++********************
**************++++++. ..........                        +++********************
***********++++++++..        ..                         .++********************
*********++++++++++...                                 .++++*******************
********++++++++++..                                   .++++*******************
*******++++++.....                                    ..++++*******************
*******+........                                     ...++++*******************
*******+... ....                                     ...++++*******************
*******+++++......                                    ..++++*******************
*******++++++++++...                                   .++++*******************
*********++++++++++...                                  ++++*******************
**********+++++++++..        ..                        ..++********************
*************++++++.. ..........                        +++********************
******************+++...+++.....                      ..+++********************
*********************+++++++++....                    ..++*********************
***********************++++++++...                     +++*********************
*************************+++++++..   .            ... .++**********************
**************************++++++++.......      ......+++***********************
****************************+++++++++....      ..++++++************************
*****************************++++++++++..     ..++++++*************************
*******************************++++++++++..  ...+++++**************************
*********************************++++++++.. ...+++++***************************
***********************************++++++....+++++*****************************
***************************************++++++++********************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
******************************************************************+++++++++++++
****************************************************************+++++++++++++++
*************************************************************++++++++++++++++++
***********************************************************++++++++++++++++++++
********************************************************+++++++++++++++++++++++
******************************************************++++++++++++++++++++++...
***************************************************+++++++++++++++++++++.......
*************************************************++++++++++++++++++++..........
***********************************************+++++++++++++++++++...       ...
********************************************++++++++++++++++++++......         
******************************************++++++++++++++++++++.......          
***************************************+++++++++++++++++++++..........         
************************************++++++++++++++++++++++...........          
********************************++++++++++++++++++++++++.........              
***************************++++++++...........+++++..............              
*********************++++++++++++....  .........................               
***************+++++++++++++++++....   .........   ............                
***********+++++++++++++++++++++.....                   ......                 
********+++++++++++++++++++++++.......                                         
******+++++++++++++++++++++++++........                                        
****+++++++++++++++++++++++++.......                                           
***+++++++++++++++++++++++.........                                            
**++++++++++++++++...........                                                  
*++++++++++++................                                                  
*++++....................                                                      
*                                                                              
*++++....................                                                      
*++++++++++++................                                                  
**++++++++++++++++...........                                                  
***+++++++++++++++++++++++.........                                            
****+++++++++++++++++++++++++.......                                           
******+++++++++++++++++++++++++........                                        
********+++++++++++++++++++++++.......                                         
***********+++++++++++++++++++++.....                   ......                 
***************+++++++++++++++++....   .........   ............                
*********************++++++++++++....  .........................               
***************************++++++++...........+++++..............              
********************************++++++++++++++++++++++++.........              
************************************++++++++++++++++++++++...........          
***************************************+++++++++++++++++++++..........         
******************************************++++++++++++++++++++.......          
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
*******************************************************************************
****************************+++++++++++++++++**********************************
***********************+++++++++++...++++++++++++******************************
********************+++++++++++++.. . .++++++++++++++**************************
*****************++++++++++++++++... ......++++++++++++************************
**************+++++++++++++++++++...   .......+++++++++++**********************
************++++++++++++++++++++....    .... ..++++++++++++********************
**********++++++++++++++++++++++......       ...++++++++++++*******************
********+++++++++++++++++++++++.......     .....++++++++++++++*****************
******++++++++++++++++++++++++.......      .....+++++++++++++++****************
****+++++++++++++++++++++++++.... .         .....+++++++++++++++***************
**+++++++++++++++++++++++++....                ...++++++++++++++++*************
*+++++++++++++++++++++++.......                ....++++++++++++++++*****
ready> 
ready> Evaluated to 0.0D0
ready> 
ready> Evaluated to 0.0D0
ready> 
ready> Evaluated to 0.0D0
ready> ; ModuleID = 'my cool jit'

declare double @printd(double)

define double @"binary:"(double %x, double %y) {
entry:
  ret double 0.000000e+00
}

define double @0() {
entry:
  %calltmp = call double @printd(double 1.230000e+02)
  %calltmp1 = call double @printd(double 4.560000e+02)
  %binop = call double @"binary:"(double %calltmp, double %calltmp1)
  %calltmp2 = call double @printd(double 7.890000e+02)
  %binop3 = call double @"binary:"(double %binop, double %calltmp2)
  ret double %binop3
}

define double @"unary!"(double %v) {
entry:
  %ifcond = fcmp ueq double %v, 0.000000e+00
  %iftmp = select i1 %ifcond, double 1.000000e+00, double 0.000000e+00
  ret double %iftmp
}

define double @unary-(double %v) {
entry:
  %subtmp = fsub double 0.000000e+00, %v
  ret double %subtmp
}

define double @"binary>"(double %LHS, double %RHS) {
entry:
  %cmptmp = fcmp ult double %RHS, %LHS
  %booltmp = uitofp i1 %cmptmp to double
  ret double %booltmp
}

define double @"binary|"(double %LHS, double %RHS) {
entry:
  %ifcond = fcmp ueq double %LHS, 0.000000e+00
  br i1 %ifcond, label %else, label %ifcont

else:                                             ; preds = %entry
  %ifcond1 = fcmp ueq double %RHS, 0.000000e+00
  %iftmp = select i1 %ifcond1, double 0.000000e+00, double 1.000000e+00
  ret double %iftmp

ifcont:                                           ; preds = %entry
  ret double 1.000000e+00
}

define double @"binary&"(double %LHS, double %RHS) {
entry:
  %unop = call double @"unary!"(double %LHS)
  %ifcond = fcmp ueq double %unop, 0.000000e+00
  br i1 %ifcond, label %else, label %ifcont

else:                                             ; preds = %entry
  %unop1 = call double @"unary!"(double %RHS)
  %unop2 = call double @"unary!"(double %unop1)
  ret double %unop2

ifcont:                                           ; preds = %entry
  ret double 0.000000e+00
}

define double @"binary="(double %LHS, double %RHS) {
entry:
  %cmptmp = fcmp ult double %LHS, %RHS
  %booltmp = uitofp i1 %cmptmp to double
  %binop = call double @"binary>"(double %LHS, double %RHS)
  %binop1 = call double @"binary|"(double %booltmp, double %binop)
  %unop = call double @"unary!"(double %binop1)
  ret double %unop
}

declare double @putchard(double)

define double @printdensity(double %d) {
entry:
  %binop = call double @"binary>"(double %d, double 8.000000e+00)
  %ifcond = fcmp ueq double %binop, 0.000000e+00
  br i1 %ifcond, label %else, label %then

then:                                             ; preds = %entry
  %calltmp = call double @putchard(double 3.200000e+01)
  ret double %calltmp

else:                                             ; preds = %entry
  %binop1 = call double @"binary>"(double %d, double 4.000000e+00)
  %ifcond2 = fcmp ueq double %binop1, 0.000000e+00
  br i1 %ifcond2, label %else4, label %then3

then3:                                            ; preds = %else
  %calltmp6 = call double @putchard(double 4.600000e+01)
  ret double %calltmp6

else4:                                            ; preds = %else
  %binop7 = call double @"binary>"(double %d, double 2.000000e+00)
  %ifcond8 = fcmp ueq double %binop7, 0.000000e+00
  br i1 %ifcond8, label %else10, label %then9

then9:                                            ; preds = %else4
  %calltmp12 = call double @putchard(double 4.300000e+01)
  ret double %calltmp12

else10:                                           ; preds = %else4
  %calltmp13 = call double @putchard(double 4.200000e+01)
  ret double %calltmp13
}

define double @1() {
entry:
  %calltmp = call double @printdensity(double 1.000000e+00)
  %calltmp1 = call double @printdensity(double 2.000000e+00)
  %binop = call double @"binary:"(double %calltmp, double %calltmp1)
  %calltmp2 = call double @printdensity(double 3.000000e+00)
  %binop3 = call double @"binary:"(double %binop, double %calltmp2)
  %calltmp4 = call double @printdensity(double 4.000000e+00)
  %binop5 = call double @"binary:"(double %binop3, double %calltmp4)
  %calltmp6 = call double @printdensity(double 5.000000e+00)
  %binop7 = call double @"binary:"(double %binop5, double %calltmp6)
  %calltmp8 = call double @printdensity(double 9.000000e+00)
  %binop9 = call double @"binary:"(double %binop7, double %calltmp8)
  %calltmp10 = call double @putchard(double 1.000000e+01)
  %binop11 = call double @"binary:"(double %binop9, double %calltmp10)
  ret double %binop11
}

define double @mandleconverger(double %real, double %imag, double %iters, double %creal, double %cimag) {
entry:
  %binop = call double @"binary>"(double %iters, double 2.550000e+02)
  %multmp = fmul double %real, %real
  %multmp1 = fmul double %imag, %imag
  %addtmp = fadd double %multmp, %multmp1
  %binop2 = call double @"binary>"(double %addtmp, double 4.000000e+00)
  %binop3 = call double @"binary|"(double %binop, double %binop2)
  %ifcond = fcmp ueq double %binop3, 0.000000e+00
  br i1 %ifcond, label %else, label %ifcont

else:                                             ; preds = %entry
  %subtmp = fsub double %multmp, %multmp1
  %addtmp6 = fadd double %subtmp, %creal
  %multmp7 = fmul double %real, 2.000000e+00
  %multmp8 = fmul double %multmp7, %imag
  %addtmp9 = fadd double %multmp8, %cimag
  %addtmp10 = fadd double %iters, 1.000000e+00
  %calltmp = call double @mandleconverger(double %addtmp6, double %addtmp9, double %addtmp10, double %creal, double %cimag)
  ret double %calltmp

ifcont:                                           ; preds = %entry
  ret double %iters
}

define double @mandleconverge(double %real, double %imag) {
entry:
  %calltmp = call double @mandleconverger(double %real, double %imag, double 0.000000e+00, double %real, double %imag)
  ret double %calltmp
}

define double @mandelhelp(double %xmin, double %xmax, double %xstep, double %ymin, double %ymax, double %ystep) {
entry:
  br label %loop

loop:                                             ; preds = %afterloop, %entry
  %y = phi double [ %ymin, %entry ], [ %nextvar4, %afterloop ]
  br label %loop1

loop1:                                            ; preds = %loop1, %loop
  %x = phi double [ %xmin, %loop ], [ %nextvar, %loop1 ]
  %calltmp = call double @mandleconverge(double %x, double %y)
  %calltmp2 = call double @printdensity(double %calltmp)
  %nextvar = fadd double %x, %xstep
  %cmptmp = fcmp ult double %x, %xmax
  br i1 %cmptmp, label %loop1, label %afterloop

afterloop:                                        ; preds = %loop1
  %calltmp3 = call double @putchard(double 1.000000e+01)
  %binop = call double @"binary:"(double 0.000000e+00, double %calltmp3)
  %nextvar4 = fadd double %y, %ystep
  %cmptmp5 = fcmp ult double %y, %ymax
  br i1 %cmptmp5, label %loop, label %afterloop8

afterloop8:                                       ; preds = %afterloop
  ret double 0.000000e+00
}

define double @mandel(double %realstart, double %imagstart, double %realmag, double %imagmag) {
entry:
  %multmp = fmul double %realmag, 7.800000e+01
  %addtmp = fadd double %multmp, %realstart
  %multmp1 = fmul double %imagmag, 4.000000e+01
  %addtmp2 = fadd double %multmp1, %imagstart
  %calltmp = call double @mandelhelp(double %realstart, double %addtmp, double %realmag, double %imagstart, double %addtmp2, double %imagmag)
  ret double %calltmp
}

define double @2() {
entry:
  %unop = call double @unary-(double 0x4002666660000000)
  %unop1 = call double @unary-(double 0x3FF4CCCCC0000000)
  %calltmp = call double @mandel(double %unop, double %unop1, double 0x3FA99999A0000000, double 0x3FB1EB8520000000)
  ret double %calltmp
}

define double @3() {
entry:
  %unop = call double @unary-(double 2.000000e+00)
  %unop1 = call double @unary-(double 1.000000e+00)
  %calltmp = call double @mandel(double %unop, double %unop1, double 0x3F947AE140000000, double 0x3FA47AE140000000)
  ret double %calltmp
}

define double @4() {
entry:
  %unop = call double @unary-(double 0x3FECCCCCC0000000)
  %unop1 = call double @unary-(double 0x3FF6666660000000)
  %calltmp = call double @mandel(double %unop, double %unop1, double 0x3F947AE140000000, double 0x3F9EB851E0000000)
  ret double %calltmp
}

